File: lib/models/question.dart
import 'enums.dart';

/// A single civics question from an official USCIS test.
class Question {
  const Question({
    required this.id,
    required this.version,
    required this.category,
    required this.section,
    required this.prompt,
    required this.answers,
    this.kind = AnswerKind.fixed,
    this.senior = false,
    this.requiredCount = 1,
    this.note,
  });

  /// Official question number within its test version (1-based).
  final int id;

  final TestVersion version;
  final QuestionCategory category;

  /// USCIS subsection, e.g. "Principles of American Democracy".
  final String section;

  final String prompt;

  /// The statically-listed accepted answers. For [AnswerKind.fixed] these are
  /// the answers; for dynamic kinds they are the official placeholder text
  /// (e.g. "Answers will vary") and the real answers are resolved at runtime.
  final List<String> answers;

  final AnswerKind kind;

  /// True if marked with an asterisk for the 65/20 exemption.
  final bool senior;

  /// How many distinct answers the applicant must give ("Name two…" -> 2).
  final int requiredCount;

  /// Extra guidance shown to the user (official notes, verify reminders, etc.).
  final String? note;

  /// A stable key used for progress/persistence.
  String get key => '${version.storageKey}_$id';

  bool get isDynamic => kind.isDynamic;
  bool get isStateDependent => kind.isStateDependent;
  bool get isTimeSensitive => kind.isTimeSensitive;
}

File: backend/src/CivicsPrep.Domain/Entities/Question.cs
using CivicsPrep.Domain.Enums;

namespace CivicsPrep.Domain.Entities;

/// <summary>A single civics question from an official USCIS test.</summary>
public class Question
{
    public int Id { get; set; }

    /// <summary>Official question number within its test version (1-based).</summary>
    public int Number { get; set; }

    public TestVersion Version { get; set; }
    public QuestionCategory Category { get; set; }

    /// <summary>USCIS subsection, e.g. "Principles of American Democracy".</summary>
    public string Section { get; set; } = string.Empty;

    public string Prompt { get; set; } = string.Empty;

    public AnswerKind Kind { get; set; }

    /// <summary>True if marked with an asterisk for the 65/20 exemption.</summary>
    public bool Senior { get; set; }

    /// <summary>How many distinct answers the applicant must give ("Name two..." -&gt; 2).</summary>
    public int RequiredCount { get; set; } = 1;

    /// <summary>Extra guidance shown to the user (official notes, verify reminders).</summary>
    public string? Note { get; set; }

    /// <summary>
    /// The statically-listed accepted answers. For dynamic kinds these hold the official
    /// placeholder text and the real answers are resolved per user at request time.
    /// </summary>
    public List<QuestionAnswer> Answers { get; set; } = [];

    public bool IsDynamic => Kind.IsDynamic();
    public bool IsStateDependent => Kind.IsStateDependent();
    public bool IsTimeSensitive => Kind.IsTimeSensitive();
}

/// <summary>One accepted answer for a <see cref="Question"/>.</summary>
public class QuestionAnswer
{
    public int Id { get; set; }
    public int QuestionId { get; set; }
    public Question? Question { get; set; }

    public string Text { get; set; } = string.Empty;

    /// <summary>Preserves the order USCIS lists the answers in.</summary>
    public int Ordinal { get; set; }
}
